pkg/platform/kube/monitoring/test/function_test.go

```diff
@@ -0,0 +1,312 @@
+				})
+			suite.Require().NoError(err, "Failed to wait for replicas to become healthy")
+
+			// wait for monitoring
+			time.Sleep(waitOnMonitoringIntervals)
+
+			// ensure function monitoring did not recover the function from its recent deploy error
+			suite.GetFunctionAndExpectState(getFunctionOptions, functionconfig.FunctionStateError)
+
+			return true
+		}, func(deployResult *platform.CreateFunctionResult) bool {
+
+			// let interval occur at least once
+			time.Sleep(waitOnMonitoringIntervals)
+
+			// ensure function in ready state, deploy passes
+			suite.GetFunctionAndExpectState(getFunctionOptions, functionconfig.FunctionStateReady)
+			return true
+		})
+}
+
+func (suite *FunctionMonitoringTestSuite) TestRecoverErrorStateFunctionWhenResourcesAvailable() {
+	functionName := "function-recovery"
+	createFunctionOptions := suite.CompileCreateFunctionOptions(functionName)
+	getFunctionOptions := &platform.GetFunctionsOptions{
+		Name:      createFunctionOptions.FunctionConfig.Meta.Name,
+		Namespace: createFunctionOptions.FunctionConfig.Meta.Namespace,
+	}
+
+	suite.DeployFunction(createFunctionOptions, func(deployResults *platform.CreateFunctionResult) bool {
+
+		// get the function
+		function := suite.GetFunction(getFunctionOptions)
+
+		// ensure function is ready
+		suite.Require().Equal(functionconfig.FunctionStateReady, function.GetStatus().State)
+
+		// get function pod, first one is enough
+		pod := suite.GetFunctionPods(functionName)[0]
+
+		// get node name on which function pod is running
+		nodeName := pod.Spec.NodeName
+
+		// mark the node as unschedulable, we want to evict the pod from there
+		suite.Logger.InfoWith("Setting cluster node as unschedulable", "nodeName", nodeName)
+		_, err := suite.KubeClientSet.CoreV1().Nodes().Update(&v1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      nodeName,
+				Namespace: suite.Namespace,
+			},
+			Spec: v1.NodeSpec{
+				Unschedulable: true,
+			},
+		})
+		suite.Require().NoError(err, "Failed to set nodes unschedulable")
+
+		// no matter how this test ends up - ensure the node is schedulable again
+		defer func() {
+			_, err := suite.KubeClientSet.CoreV1().Nodes().Update(&v1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      nodeName,
+					Namespace: suite.Namespace,
+				},
+				Spec: v1.NodeSpec{
+					Unschedulable: false,
+				},
+			})
+			suite.Require().NoError(err)
+		}()
+
+		// delete function pod
+		zeroSeconds := int64(0)
+		suite.Logger.InfoWith("Deleting function pod", "podName", pod.Name)
+		err = suite.KubeClientSet.CoreV1().Pods(suite.Namespace).Delete(pod.Name,
+			&metav1.DeleteOptions{
+				GracePeriodSeconds: &zeroSeconds,
+			})
+		suite.Require().NoError(err, "Failed to delete function pod")
+
+		// wait for controller to mark function in error due to pods are unschedulable
+		err = common.RetryUntilSuccessful(2*suite.Controller.GetFunctionMonitoringInterval(),
```

Comment:
there is `waitForFunctionState`

Reply:
it is part of nuctl test suite, but good point, I will add one to kube test suite.